posts/2024-11-01-gsoc-2024-results.md

```diff
@@ -0,0 +1,180 @@
+
+Even though this project did not fulfill its original goals and more work will be needed in the future to get a functional prototype of WebAssembly proc macros, we are thankful for Apurva's contributions. The WebAssembly loading prototype is a good start, and Apurva's exploration of proc macro internals should serve as a useful reference for anyone working on this feature in the future. Going forward, we will try to describe more incremental steps for our GSoC projects, as this project was perhaps too ambitious from the start.
+
+### Tokio async support in Miri
+
+- Contributor: [Tiffany Pek Yuan](https://github.com/tiif)
+- Mentor: [Oli Scherer](https://github.com/oli-obk)
+- [Final report](https://gist.github.com/tiif/3e08ba6e8cfb1d078e6155410108ae48)
+
+[miri][miri] is an intepreter that can find possible instances of undefined behavior in your Rust code, which is being used across the Rust ecosystem. However, it was not possible to run it on any non-trivial programs (those that ever `await` on anything) using [tokio][tokio], due a to a fundamental missing feature: support for the `epoll` syscall on Linux (and similar APIs on other major platforms).
+
+Tiffany implemented the basic `epoll` operations needed to cover the majority of the tokio test suite, by crafting pure `libc` code examples that exercised those `epoll` operations, and then implementing their emulation in miri itself. At times, this required refactoring core miri components like file description handling, as they were originally created solely with blocking operations in mind.
+
+Suprising to everyone (though probably not tokio-internals experts), once these core `epoll` operations were finished, operations like async file reading and writing started working out of the box! Due to limitations of non-blocking file operations offered by operating systems, tokio wrapped those file operations in dedicated threads, which was already supported by miri.
+
+Once Tiffany has finished the project, including stretch goals like implementing async file operations, she proceeded to contact tokio maintainers and worked with them to run miri on most tokio tests in CI. And we have good news: so far no soundness problems have been discovered! Tiffany has become a regular contributor to miri, focusing on continuing to expand the set of supported non-blocking file descriptor operations. We thank her for all her contributions!
```

Comment:
Thanks for the writeup, just a few nitpicks:

I remember file description handling is refactored to support blocking operation for eventfd and friends. 

```suggestion
Tiffany implemented the basic `epoll` operations needed to cover the majority of the tokio test suite, by crafting pure `libc` code examples that exercised those `epoll` operations, and then implementing their emulation in miri itself. At times, this required refactoring core miri components like file description handling, as they were originally created solely with non-blocking operations in mind.

Suprising to everyone (though probably not tokio-internals experts), once these core `epoll` operations were finished, operations like async file reading and writing started working out of the box! Due to limitations of non-blocking file operations offered by operating systems, tokio wrapped those file operations in dedicated threads, which was already supported by miri.

Once Tiffany has finished the project, including stretch goals like implementing async file operations, she proceeded to contact tokio maintainers and worked with them to run miri on most tokio tests in CI. And we have good news: so far no soundness problems have been discovered! Tiffany has become a regular contributor to miri, focusing on continuing to expand the set of supported blocking file descriptor operations. We thank her for all her contributions!
```

Reply:
Thanks! Did some modifications.